File: workers/packages/engine/src/splits.ts
import { EngineError } from './errors.js';
import {
  BP_TOTAL,
  MAX_AMOUNT_PAISE,
  MAX_PARTICIPANTS,
  type SplitInput,
  type SplitParticipant,
} from './types.js';

/**
 * Compute per-participant shares for an expense (M8).
 *
 * One deterministic rounding rule for equal/percent/shares:
 * largest-remainder, ties broken by user_id ascending. Exact splits must
 * sum precisely or are rejected — never auto-fixed.
 *
 * Postcondition (M7): Σ(shares) == amountTotal, every share >= 0.
 */
export function computeShares(input: SplitInput): Map<string, number> {
  const { splitType, amountTotal, participants } = input;

  if (!Number.isInteger(amountTotal) || amountTotal <= 0) {
    throw new EngineError('invalid_amount');
  }
  if (amountTotal > MAX_AMOUNT_PAISE) {
    throw new EngineError('amount_out_of_bounds');
  }
  if (participants.length === 0) {
    throw new EngineError('no_participants');
  }
  if (participants.length > MAX_PARTICIPANTS) {
    throw new EngineError('too_many_participants');
  }
  const ids = new Set<string>();
  for (const p of participants) {
    if (ids.has(p.userId)) throw new EngineError('duplicate_participant', p.userId);
    ids.add(p.userId);
  }

  switch (splitType) {
    case 'exact':
      return exactShares(amountTotal, participants);
    case 'equal':
      return largestRemainder(
        amountTotal,
        participants.map((p) => ({ userId: p.userId, numerator: 1n })),
        BigInt(participants.length),
      );
    case 'percent': {
      let sum = 0;
      for (const p of participants) {
        if (p.bp === undefined || !Number.isInteger(p.bp) || p.bp < 0 || p.bp > BP_TOTAL) {
          throw new EngineError('invalid_percent', p.userId);
        }
        sum += p.bp;
      }
      if (sum !== BP_TOTAL) throw new EngineError('percent_sum_mismatch');
      return largestRemainder(
        amountTotal,
        participants.map((p) => ({ userId: p.userId, numerator: BigInt(p.bp!) })),
        BigInt(BP_TOTAL),
      );
    }
    case 'shares': {
      let totalWeight = 0n;
      for (const p of participants) {
        if (p.weight === undefined || !Number.isInteger(p.weight) || p.weight < 1) {
          throw new EngineError('invalid_weight', p.userId);
        }
        totalWeight += BigInt(p.weight);
      }
      return largestRemainder(
        amountTotal,
        participants.map((p) => ({ userId: p.userId, numerator: BigInt(p.weight!) })),
        totalWeight,
      );
    }
  }
}

function exactShares(amountTotal: number, participants: SplitParticipant[]): Map<string, number> {
  let sum = 0;
  for (const p of participants) {
    if (p.amount === undefined || !Number.isInteger(p.amount)) {
      throw new EngineError('negative_share', p.userId);
    }
    if (p.amount < 0) throw new EngineError('negative_share', p.userId);
    sum += p.amount;
  }
  if (sum !== amountTotal) throw new EngineError('split_sum_mismatch');
  return new Map(participants.map((p) => [p.userId, p.amount!]));
}

/**
 * Largest-remainder apportionment over ideal shares total*numerator/denominator.
 * BigInt throughout — total*numerator can exceed 2^53 (D2 bounds make the
 * inputs legal up to 10^12 * 10^4 = 10^16).
 */
function largestRemainder(
  total: number,
  entries: Array<{ userId: string; numerator: bigint }>,
  denominator: bigint,
): Map<string, number> {
  const totalBig = BigInt(total);
  const computed = entries.map((e) => {
    const product = totalBig * e.numerator;
    return {
      userId: e.userId,
      floor: Number(product / denominator),
      remainder: product % denominator,
    };
  });

  let leftover = total - computed.reduce((acc, c) => acc + c.floor, 0);

  const order = [...computed].sort((x, y) => {
    if (x.remainder !== y.remainder) return x.remainder > y.remainder ? -1 : 1;
    return x.userId < y.userId ? -1 : x.userId > y.userId ? 1 : 0;
  });

  const bonus = new Map<string, number>();
  for (const entry of order) {
    if (leftover === 0) break;
    bonus.set(entry.userId, 1);
    leftover -= 1;
  }

  const shares = new Map<string, number>();
  for (const c of computed) {
    shares.set(c.userId, c.floor + (bonus.get(c.userId) ?? 0));
  }

  // M7 postcondition — belt and braces.
  let sum = 0;
  for (const v of shares.values()) sum += v;
  if (sum !== total) throw new EngineError('split_sum_mismatch', 'internal rounding error');
  return shares;
}

File: app/lib/engine/splits.dart
import 'errors.dart';
import 'types.dart';

/// Compute per-participant shares for an expense (M8).
///
/// One deterministic rounding rule for equal/percent/shares:
/// largest-remainder, ties broken by user_id ascending. Exact splits must
/// sum precisely or are rejected — never auto-fixed.
///
/// Postcondition (M7): Σ(shares) == amountTotal, every share >= 0.
/// Dart ints are 64-bit: total * bp tops out at 10^16, well inside range
/// (the TS engine uses BigInt for the same product).
Map<String, int> computeShares({
  required String splitType,
  required int amountTotal,
  required List<SplitParticipant> participants,
}) {
  if (amountTotal <= 0) throw const EngineError(EngineError.invalidAmount);
  if (amountTotal > maxAmountPaise) {
    throw const EngineError(EngineError.amountOutOfBounds);
  }
  if (participants.isEmpty) throw const EngineError(EngineError.noParticipants);
  if (participants.length > maxParticipants) {
    throw const EngineError(EngineError.tooManyParticipants);
  }
  final ids = <String>{};
  for (final p in participants) {
    if (!ids.add(p.userId)) {
      throw EngineError(EngineError.duplicateParticipant, p.userId);
    }
  }

  switch (splitType) {
    case 'exact':
      return _exactShares(amountTotal, participants);
    case 'equal':
      return _largestRemainder(
        amountTotal,
        [for (final p in participants) (userId: p.userId, numerator: 1)],
        participants.length,
      );
    case 'percent':
      var sum = 0;
      for (final p in participants) {
        final bp = p.bp;
        if (bp == null || bp < 0 || bp > bpTotal) {
          throw EngineError(EngineError.invalidPercent, p.userId);
        }
        sum += bp;
      }
      if (sum != bpTotal) throw const EngineError(EngineError.percentSumMismatch);
      return _largestRemainder(
        amountTotal,
        [for (final p in participants) (userId: p.userId, numerator: p.bp!)],
        bpTotal,
      );
    case 'shares':
      var totalWeight = 0;
      for (final p in participants) {
        final weight = p.weight;
        if (weight == null || weight < 1) {
          throw EngineError(EngineError.invalidWeight, p.userId);
        }
        totalWeight += weight;
      }
      return _largestRemainder(
        amountTotal,
        [for (final p in participants) (userId: p.userId, numerator: p.weight!)],
        totalWeight,
      );
    default:
      throw ArgumentError.value(splitType, 'splitType');
  }
}

Map<String, int> _exactShares(int amountTotal, List<SplitParticipant> participants) {
  var sum = 0;
  for (final p in participants) {
    final amount = p.amount;
    if (amount == null || amount < 0) {
      throw EngineError(EngineError.negativeShare, p.userId);
    }
    sum += amount;
  }
  if (sum != amountTotal) throw const EngineError(EngineError.splitSumMismatch);
  return {for (final p in participants) p.userId: p.amount!};
}

/// Largest-remainder apportionment over ideal shares total*numerator/denominator.
Map<String, int> _largestRemainder(
  int total,
  List<({String userId, int numerator})> entries,
  int denominator,
) {
  final computed = [
    for (final e in entries)
      (
        userId: e.userId,
        floor: (total * e.numerator) ~/ denominator,
        remainder: (total * e.numerator) % denominator,
      ),
  ];

  var leftover = total - computed.fold<int>(0, (acc, c) => acc + c.floor);

  final order = [...computed]..sort((x, y) {
    if (x.remainder != y.remainder) return y.remainder.compareTo(x.remainder);
    return x.userId.compareTo(y.userId);
  });

  final bonus = <String, int>{};
  for (final entry in order) {
    if (leftover == 0) break;
    bonus[entry.userId] = 1;
    leftover -= 1;
  }

  final shares = <String, int>{
    for (final c in computed) c.userId: c.floor + (bonus[c.userId] ?? 0),
  };

  // M7 postcondition — belt and braces.
  final sum = shares.values.fold<int>(0, (acc, v) => acc + v);
  if (sum != total) {
    throw const EngineError(EngineError.splitSumMismatch, 'internal rounding error');
  }
  return shares;
}
